From 16e2936858ce15ba0b2ebe570616b9e8574648ac Mon Sep 17 00:00:00 2001
From: Karthik D A <karthikda@maxlinear.com>
Date: Tue, 1 Jul 2025 22:44:07 +0800
Subject: [PATCH] WLANRTSYS-88975 iwlwav-hostap-ng DL SCS support

Description: Implementation of the Service Classified Stream- SCS feature.
The iwlwav-hostapd-ng must handle and parse the SCS Add, Remove and Change request.
All the changes are under the CONFIG_SCS_MXL flag.

Signed-off-by: Karthik D A <karthikda@maxlinear.com>
---
 hostapd/mxl_config.c             |  10 +
 hostapd/mxl_ctrl_iface.c         |  33 ++++
 hostapd/mxl_hostapd_cli.c        |   5 +
 hostapd/mxl_hostapd_cli.h        |   4 +
 src/ap/mxl_config.h              |   3 +
 src/ap/mxl_hostapd.c             |   6 +
 src/ap/mxl_wmm.c                 | 318 +++++++++++++++++++++++++------
 src/ap/mxl_wmm.h                 |   2 +
 src/common/mxl_ieee802_11_defs.h |  19 ++
 src/common/vendor_cmds_copy.h    |  57 ++++++
 10 files changed, 402 insertions(+), 55 deletions(-)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index 5ad436640..cdd1ba64b 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -2010,6 +2010,16 @@ int mxl_hostapd_config_fill (struct hostapd_config *conf,
 		}
 		mxl_bss_conf->vendor_4addr = val;
 #endif /* CONFIG_LEGACY_MULTI_AP_MXL */
+#ifdef CONFIG_SCS_MXL
+	} else if (os_strcmp(buf, "scs_enable") == 0) {
+		int val = atoi(pos);
+		if (!IS_VALID_BOOL_RANGE(val)) {
+			wpa_printf(MSG_ERROR, "Line %d: invalid val %d",
+					line, val);
+			return 0;
+		}
+		mxl_bss_conf->scs_enable = val;
+#endif /* CONFIG_SCS_MXL */
 	} else
 		return 0;
 
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 28d2b4136..72c0c79b6 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -3109,6 +3109,34 @@ int mxl_hostapd_ctrl_iface_update_mbss_beacon(struct hostapd_data *hapd)
 	return 0;
 }
 
+#ifdef CONFIG_SCS_MXL
+static int mxl_hostapd_drv_get_scs_enable(struct hostapd_data *hapd, int *enabled)
+{
+	int ret;
+
+	ret = MXL_DRV_VENDOR_GET(LTQ_NL80211_VENDOR_SUBCMD_GET_SCS_ENABLE, NULL, 0, enabled, sizeof(*enabled));
+	if (ret < 0)
+		wpa_printf(MSG_ERROR, "nl80211: sending GET_SCS_ENABLE failed: %i (%s)", ret, strerror(-ret));
+	return ret;
+}
+
+static int mxl_hostapd_ctrl_iface_get_scs_enable(struct hostapd_data *hapd, char *reply, size_t reply_size)
+{
+	int res = 0;
+	int enabled;
+
+	if (mxl_hostapd_drv_get_scs_enable(hapd, &enabled))
+		return -1;
+
+	res = sprintf_s(reply, reply_size, "%d\n", enabled);
+	if (res <= 0) {
+		wpa_printf(MSG_ERROR, "%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return res;
+}
+#endif
 
 int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 					      char *buf, char *reply,
@@ -3342,6 +3370,11 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "BSS_RECONF", 10) == 0) {
 		if (mxl_hostapd_ctrl_iface_bss_reconf(hapd->iface))
 			*reply_len = -1;
+#ifdef CONFIG_SCS_MXL
+	} else if (os_strncmp(buf, "GET_SCS_ENABLE", 14) == 0) {
+		if (mxl_hostapd_ctrl_iface_get_scs_enable(hapd, reply, reply_size) < 0)
+			*reply_len = -1;
+#endif /* CONFIG_SCS */
 	} else {
 		return 0;
 	}
diff --git a/hostapd/mxl_hostapd_cli.c b/hostapd/mxl_hostapd_cli.c
index 75e1e1108..70689e2ee 100644
--- a/hostapd/mxl_hostapd_cli.c
+++ b/hostapd/mxl_hostapd_cli.c
@@ -559,3 +559,8 @@ int mxl_hostapd_cli_cmd_bss_reconf(struct wpa_ctrl *ctrl, int argc, char *argv[]
 {
 	return mxl_wpa_ctrl_command(ctrl, "BSS_RECONF");
 }
+
+int mxl_hostapd_cli_cmd_get_scs_enable(struct wpa_ctrl *ctrl, int argc, char *argv[])
+{
+	return mxl_wpa_ctrl_command(ctrl, "GET_SCS_ENABLE");
+}
diff --git a/hostapd/mxl_hostapd_cli.h b/hostapd/mxl_hostapd_cli.h
index bb64108a0..dd714f440 100644
--- a/hostapd/mxl_hostapd_cli.h
+++ b/hostapd/mxl_hostapd_cli.h
@@ -108,6 +108,8 @@ int mxl_wpa_ctrl_command(struct wpa_ctrl *ctrl, const char *cmd);
 	  MXL_CLI_BSS_NAME_PREFIX "= get assocresp elements" }, \
 	{ "bss_reconf", mxl_hostapd_cli_cmd_bss_reconf, NULL, \
 	  " = do reconf on all changed bss in radio"}, \
+	{ "get_scs_enable", mxl_hostapd_cli_cmd_get_scs_enable, NULL, \
+	  " = Get scs enable/disable status"}, \
 
 #ifdef CONFIG_ACS
 #define MXL_CLI_COMMANDS_ACS \
@@ -229,5 +231,7 @@ int mxl_hostapd_cli_cmd_get_authresp_elements(struct wpa_ctrl *ctrl, int argc, c
 int mxl_hostapd_cli_cmd_set_assocresp_elements(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_get_assocresp_elements(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_bss_reconf(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_get_scs_enable(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+
 
 #endif /* MXL_HOSTAPD_CLI_H */
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index 859013e49..4b6f49ee9 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -524,6 +524,9 @@ struct mxl_hostapd_bss_config {
 #ifdef CONFIG_LEGACY_MULTI_AP_MXL
 	bool vendor_4addr;
 #endif /* CONFIG_LEGACY_MULTI_AP_MXL */
+#ifdef CONFIG_SCS_MXL
+	int scs_enable; /* CONFIG_SCS_MXL */
+#endif
 };
 
 int mxl_hostapd_config_fill (struct hostapd_config *conf,
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 9082222c1..37c058574 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -1745,6 +1745,12 @@ int mxl_hostapd_setup_bss_pre(struct hostapd_data *hapd)
 			(u8*)&hapd->conf->mxl_bss_conf.sDisableSoftblock,
 			sizeof(hapd->conf->mxl_bss_conf.sDisableSoftblock)))
 		wpa_printf(MSG_WARNING, "set sDisableSoftblock failed");
+#ifdef CONFIG_SCS_MXL
+	if (MXL_DRV_VENDOR_SET(LTQ_NL80211_VENDOR_SUBCMD_SET_SCS_ENABLE,
+			(u8*)&hapd->conf->mxl_bss_conf.scs_enable,
+			sizeof(hapd->conf->mxl_bss_conf.scs_enable)))
+		wpa_printf(MSG_WARNING, "Enable SCS Failed");
+#endif
 
 	return 0;
 }
diff --git a/src/ap/mxl_wmm.c b/src/ap/mxl_wmm.c
index 30b6cbe2b..03f05465d 100644
--- a/src/ap/mxl_wmm.c
+++ b/src/ap/mxl_wmm.c
@@ -204,16 +204,228 @@ static void mxl_send_scs_resp(struct hostapd_data *hapd,
 		wpa_printf(MSG_DEBUG, "ROBUST AV scs_action_send_resp: send failed");
 }
 
+int mxl_scs_parse_qos_characteristics_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req)
+{
+	u32 ctrl_info;
+
+	if (*pos++ != WLAN_EID_EXT_QOS_CHARACTERISTICS)
+		return WLAN_STATUS_UNSPECIFIED_FAILURE;
+
+	os_memcpy(&ctrl_info, pos, sizeof(u32));
+	pos += sizeof(u32); /* Length of control info */
+
+	os_memcpy(&(scs_add_req->minServiceInterval), pos, sizeof(u32));
+	pos += sizeof(u32); /* Length of minimum service time */
+
+	os_memcpy(&(scs_add_req->maxServiceInterval), pos, sizeof(u32));
+	pos += sizeof(u32); /* Length of maximum service time */
+
+	os_memcpy(&(scs_add_req->minDataRate), pos, 3 * sizeof(u8));
+	pos += 3 * sizeof(u8); /* Length of min data rate */
+
+	os_memcpy(&(scs_add_req->delayBound), pos, 3 * sizeof(u8));
+	pos += 3 * sizeof(u8); /* Length of delay bound */
+
+	scs_add_req->direction = QOS_CTRL_INFO_DIR_MASK & ctrl_info;
+	scs_add_req->tid = QOS_CTRL_INFO_TID_MASK  &  ( ctrl_info >> QOS_CTRL_INFO_TID_SHIFT);
+	scs_add_req->userPrio = QOS_CTRL_INFO_USER_PRIORITY_MASK &
+		(ctrl_info >> QOS_CTRL_INFO_USER_PRIORITY_SHIFT);
+	scs_add_req->bitmapNextParamPresence  = QOS_CTRL_INFO_BIT_MAP_MASK  &
+		(ctrl_info >> QOS_CTRL_INFO_BIT_MAP_SHIFT);
+	scs_add_req->linkId = QOS_CTRL_INFO_LINKID_MASK &
+		(ctrl_info >> QOS_CTRL_INFO_LINKID_SHIFT);
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MAX_MSDU_SIZE)) {
+		os_memcpy(&(scs_add_req->maxMsduSize), pos, sizeof(u16));
+		pos += sizeof(u16); /* Length of maxMsduSize */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_SERVICE_START_TIME)) {
+		os_memcpy(&(scs_add_req->serviceStartTime), pos, sizeof(u32));
+		pos += sizeof(u32); /* Length of serviceStartTime */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_SERVICE_START_TIME_LINK_ID)) {
+		os_memcpy(&(scs_add_req->serviceStartTimeLinkId), pos, sizeof(u8));
+		pos += sizeof(u8); /* Length of serviceStartTimeLinkId */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MEAN_DATA_RATE)) {
+		os_memcpy(&(scs_add_req->meanDataRate), pos, 3 * sizeof(u8));
+		pos += 3 * sizeof(u8); /* Length of meanDataRate */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_BURST_SIZE)) {
+		os_memcpy(&(scs_add_req->burstSize), pos, sizeof(u32));
+		pos += sizeof(u32); /* Length of burstSize */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MSDU_LIFE_TIME)) {
+		os_memcpy(&(scs_add_req->msduLifeTime), pos, sizeof(u16));
+		pos += sizeof(u16); /* length of msduLifeTime */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MSDU_DELIVERY_RATIO)) {
+		os_memcpy(&(scs_add_req->msduDeliveryRatio), pos, sizeof(u8));
+		pos += sizeof(u8); /* Length of msduDeliveryRatio */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MSDU_COUNT_EXPONENT)) {
+		os_memcpy(&(scs_add_req->msduCountExponent), pos, sizeof(u8));
+		pos += sizeof(u8); /* Length of msduCountExponent */
+	}
+
+	if (scs_add_req->bitmapNextParamPresence & BIT(QOS_MEDIUM_TIME)) {
+		os_memcpy(&(scs_add_req->mediumTime), pos, sizeof(u16));
+		pos += sizeof(u16); /* length of mediumTime */
+	}
+
+	return WLAN_STATUS_SUCCESS;
+}
+
+int mxl_scs_parse_tclas_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req, size_t len, u8 *elems_num)
+{
+	const u8 *end;
+	u8 class_type, version;
+	u8 mask = 0, tclas_len = 0;
+	bool has_type4 = false, has_type10 = false;
+
+	tclas_len = len;
+	scs_add_req->tclasInfo.tclasLen = tclas_len;
+	scs_add_req->tclasInfo.tclasElemsUp = *pos++; /* populate user priority */
+	len--;
+
+	end = pos + len;
+
+	while (pos < end) {
+		class_type = *pos++;
+		switch (class_type) { /* Classifier type 4 and 10 are allowed */
+			case SCS_CLASSIFIER_TYPE4:
+				mask = *pos++;
+
+				/* Process classifiers frames */
+				if (mask & BIT(SCS_CLASS_TYPE4_VERSION))
+					scs_add_req->type4Params.version = *pos++; /* Version */
+
+				if (scs_add_req->type4Params.version != IPV4 && scs_add_req->type4Params.version != IPV6)
+					return WLAN_STATUS_UNSPECIFIED_FAILURE;
+
+				version = scs_add_req->type4Params.version;
+
+				if (mask & BIT(SCS_CLASS_TYPE4_SRC_IP)) { /* source IP address */
+					if (version == IPV4) {
+						memcpy(&(scs_add_req->type4Params.u.v4.srcIp), pos, sizeof(u32));
+						pos += sizeof(u32);
+					} else {
+						memcpy(scs_add_req->type4Params.u.v6.srcIp, pos, 16);
+						pos += 16;
+					}
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_DST_IP)) { /* Destination IP address */
+					if (version == IPV4) {
+						memcpy(&(scs_add_req->type4Params.u.v4.dstIp), pos, sizeof(u32));
+						pos += sizeof(u32);
+					} else {
+						memcpy(scs_add_req->type4Params.u.v6.dstIp, pos, 16);
+						pos += 16;
+					}
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_SRC_PORT)) { /* Source port */
+					if (version == IPV4) {
+						memcpy(&(scs_add_req->type4Params.u.v4.srcPort), pos, sizeof(u16));
+						pos += sizeof(u16);
+					} else {
+						memcpy(&(scs_add_req->type4Params.u.v6.srcPort), pos, sizeof(u16));
+						pos += sizeof(u16);
+					}
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_DST_PORT)) { /* Dst port */
+					if (version == IPV4) {
+						memcpy(&(scs_add_req->type4Params.u.v4.dstPort), pos, sizeof(u16));
+						pos += sizeof(u16);
+					} else {
+						memcpy(&(scs_add_req->type4Params.u.v6.dstPort), pos, sizeof(u16));
+						pos += sizeof(u16);
+					}
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_DSCP)) { /* DSCP */
+					if (version == IPV4)
+						scs_add_req->type4Params.u.v4.dscp = *pos++;
+					else
+						scs_add_req->type4Params.u.v6.dscp = *pos++;
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_PROTO)) { /* Protocol or next header */
+					if (version == IPV4)
+						scs_add_req->type4Params.u.v4.proto = *pos++;
+					else
+						scs_add_req->type4Params.u.v6.nextHeader = *pos++;
+				}
+
+				if (mask & BIT(SCS_CLASS_TYPE4_FLOW_LABEL)) { /* Flow label only for ipv6 */
+						memcpy(scs_add_req->type4Params.u.v6.flowLabel, pos, 3);
+						pos += 3;
+				}
+
+				has_type4 = true;
+				scs_add_req->tclasInfo.tclasElemstype = 4;
+				elems_num ++;
+				break;
+			 case SCS_CLASSIFIER_TYPE10:
+				scs_add_req->type10Params.protoInstance = *pos++;
+				scs_add_req->type10Params.protoNum = *pos++;
+				scs_add_req->type10Params.filterLen = (tclas_len - 4) / 2;
+				memcpy(scs_add_req->type10Params.filterValue, pos, scs_add_req->type10Params.filterLen);
+				pos += scs_add_req->type10Params.filterLen;
+				memcpy(scs_add_req->type10Params.filterMask, pos, scs_add_req->type10Params.filterLen);
+				pos += scs_add_req->type10Params.filterLen;
+				has_type10 = true;
+				elems_num ++;
+				break;
+			 default:
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		}
+	}
+	/* TCLAS type 10 without TCLAS type 4, one of the illegal class options.
+	 * Not sure on the order of class types , in case if its random i.e class 10
+	 * appears before class 4. So, has_type10 and has_type6 variables are used *
+	 * and the tclas type 10 without tclas 4 is checked after the loop ends */
+	if (has_type10) {
+		if (!has_type4)
+			return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		/* TCLAS type 10 with a protocol different than 17 or 50 is illegal */
+		if (scs_add_req->type10Params.protoInstance != 0 ||
+		    (scs_add_req->type10Params.protoNum != 17 &&
+		    scs_add_req->type10Params.protoNum != 50))
+			return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		if (version == IPV4) {
+			if (scs_add_req->type4Params.u.v4.srcPort != 0)
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		}
+		else if (version == IPV6) {
+			if (scs_add_req->type4Params.u.v6.srcPort != 0)
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		}
+		scs_add_req->tclasInfo.tclasElemstype = 10;
+	}
+	return WLAN_STATUS_SUCCESS;
+}
+
 int mxl_scs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len)
 {
-	const u8 *pos,*pos1;
-	pos = mgmt->u.action.u.scs_action_req.variable;
-	struct qos_characteristic_element *q = (struct qos_characteristic_element *) pos;
+	const u8 *pos, *end;
+	u8 eid, elen;
 	struct sta_info *sta;
 	struct mxl_scs_add_req scs_add_req;
-	os_memset(&scs_add_req, 0, sizeof(struct mxl_scs_add_req));
+	pos = mgmt->u.action.u.scs_action_req.variable;
+	u8 tclas_elems_num = 0;
+	bool has_tclas_elems = false, has_tclas_process = false;
 
-	wpa_printf(MSG_DEBUG,"SCS: eid=%d,length=%d,eid_ext=%d",q->eid,q->length,q->eid_ext);
+	os_memset(&scs_add_req, 0, sizeof(struct mxl_scs_add_req));
 
 	sta = ap_get_sta(hapd, mgmt->sa);
 	if (sta == NULL) {
@@ -226,68 +438,57 @@ int mxl_scs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
 	}
 
+	end = (const u8 *) mgmt + len;
 	scs_add_req.aid = sta->aid;
 	scs_add_req.status = 0;
 	scs_add_req.scsid = mgmt->u.action.u.scs_action_req.scsId;
-	scs_add_req.minDataRate = q->minDataRate[0] | (q->minDataRate[1] << 8) | (q->minDataRate[2] << 16);
-	scs_add_req.delayBound = q->delayBound[0] | (q->delayBound[1] << 8) | (q->delayBound[2] << 16);
-	u32 ctrl_info = le_to_host32(q->ctrl_info);
-	scs_add_req.minServiceInterval = le_to_host32(q->minServiceInterval);
-	scs_add_req.maxServiceInterval = le_to_host32(q->maxServiceInterval);
-	scs_add_req.tid = QOS_CTRL_INFO_TID_MASK  &  ( ctrl_info >> QOS_CTRL_INFO_TID_SHIFT);
-	scs_add_req.userPrio = QOS_CTRL_INFO_USER_PRIORITY_MASK & (ctrl_info >> QOS_CTRL_INFO_USER_PRIORITY_SHIFT);
-	scs_add_req.bitmapNextParamPresence  = QOS_CTRL_INFO_BIT_MAP_MASK  & ( ctrl_info >> QOS_CTRL_INFO_BIT_MAP_SHIFT);
-	scs_add_req.linkId = QOS_CTRL_INFO_LINKID_MASK & ( ctrl_info >> QOS_CTRL_INFO_LINKID_SHIFT);
-
-	pos1 = q->variable;
-
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MAX_MSDU_SIZE)) {
-		os_memcpy(&(scs_add_req.maxMsduSize), pos1, sizeof(u16));
-		pos1 += sizeof(u16); // length of maxMsduSize
-	}
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_SERVICE_START_TIME)) {
-		os_memcpy(&(scs_add_req.serviceStartTime), pos1, sizeof(u32));
-		pos1 += sizeof(u32); // length of serviceStartTime
-	}
-
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_SERVICE_START_TIME_LINK_ID)) {
-		os_memcpy(&(scs_add_req.serviceStartTimeLinkId), pos1, sizeof(u8));
-		pos1 += sizeof(u8); // length of serviceStartTimeLinkId
-	}
+	while (pos + 2 <= end) {
+		eid = *pos++;
+		elen = *pos++;
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MEAN_DATA_RATE)) {
-		os_memcpy(&(scs_add_req.meanDataRate), pos1, 3 * sizeof(u8));
-		pos1 += 3 * sizeof(u8); // length of meanDataRate
-	}
+		if (end - pos < elen)
+			return WLAN_STATUS_UNSPECIFIED_FAILURE;
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_BURST_SIZE)) {
-		os_memcpy(&(scs_add_req.burstSize), pos1, sizeof(u32));
-		pos1 += sizeof(u32); // length of burstSize
-	}
-
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MSDU_LIFE_TIME)) {
-		os_memcpy(&(scs_add_req.msduLifeTime), pos1, sizeof(u16));
-		pos1 += sizeof(u16); // length of msduLifeTime
-	}
+		switch (eid) {
+		case WLAN_EID_INTRA_ACCESS_CATEGORY_PRIORITY:
+			scs_add_req.intraAccessPriority = *pos;
+			break;
+		case WLAN_EID_TCLAS:
+			if (mxl_scs_parse_tclas_elems(pos, &scs_add_req, elen, &tclas_elems_num))
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+			has_tclas_elems = true;
+			break;
+		case WLAN_EID_TCLAS_PROCESSING:
+			/* the processing field = 0 , indicates it should match to *
+			 * all the parameters associated with TCLAS elems */
+			if (*pos != 0 || tclas_elems_num <=1)
+			   return WLAN_STATUS_UNSPECIFIED_FAILURE;
+			has_tclas_process = true;
+			break;
+		case WLAN_EID_EXTENSION:
+			if (mxl_scs_parse_qos_characteristics_elems(pos, &scs_add_req))
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+			break;
+		case WLAN_EID_VENDOR_SPECIFIC:
+			/*Need to verify, EID populated as optional subelement *
+			 * Table 9-287—Optional subelement IDs for SCS Descriptor element *
+			 * At present reject vendor specific SCS request cases */
+			return WLAN_STATUS_UNSPECIFIED_FAILURE;
+		default:
+		     return WLAN_STATUS_UNSPECIFIED_FAILURE;
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MSDU_DELIVERY_RATIO)) {
-		os_memcpy(&(scs_add_req.msduDeliveryRatio), pos1, sizeof(u8));
-		pos1 += sizeof(u8); // length of msduDeliveryRatio
-	}
+		}
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MSDU_COUNT_EXPONENT)) {
-		os_memcpy(&(scs_add_req.msduCountExponent), pos1, sizeof(u8));
-		pos1 += sizeof(u8); // length of msduCountExponent
+		pos += elen;
 	}
 
-	if (scs_add_req.bitmapNextParamPresence & BIT(QOS_MEDIUM_TIME)) {
-		os_memcpy(&(scs_add_req.mediumTime), pos1, sizeof(u16));
-		pos1 += sizeof(u16); // length of mediumTime
-	}
+	/* TCLAS elems without TCLAS_PROCESSING elems is illegal */
+	if (has_tclas_elems && !has_tclas_process)
+		return WLAN_STATUS_UNSPECIFIED_FAILURE;
 
 	if (MXL_DRV_VENDOR_SET(LTQ_NL80211_VENDOR_SUBCMD_SCS_ADD_REQ,
-	    &scs_add_req, sizeof(struct mxl_scs_add_req))) {
+				&scs_add_req, sizeof(struct mxl_scs_add_req))) {
 		wpa_printf(MSG_ERROR, "send LTQ_NL80211_VENDOR_SUBCMD_SEND_SCS_ADD_REQ failed!!!");
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
 	}
@@ -347,6 +548,13 @@ void mxl_handle_scs_req(struct hostapd_data *hapd,
 			status = mxl_scs_remove_req(hapd, mgmt, len);
 			mxl_send_scs_resp(hapd, mgmt, status);
 			break;
+		case SCS_REQ_CHANGE:
+			status = mxl_scs_remove_req(hapd, mgmt, len);
+			if (!status)
+				mxl_send_scs_resp(hapd, mgmt, status);
+			status = mxl_scs_add_req(hapd, mgmt,len);
+			mxl_send_scs_resp(hapd, mgmt, status);
+			break;
 		default:
 			hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
 			        HOSTAPD_LEVEL_DEBUG,
diff --git a/src/ap/mxl_wmm.h b/src/ap/mxl_wmm.h
index 1e3d0ed3b..79e5fcb0f 100644
--- a/src/ap/mxl_wmm.h
+++ b/src/ap/mxl_wmm.h
@@ -29,6 +29,8 @@ void mxl_hostapd_scs_action(struct hostapd_data *hapd, const struct ieee80211_mg
 void mxl_handle_scs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
 int mxl_scs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
 int mxl_scs_remove_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
+int mxl_scs_parse_qos_characteristics_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req);
+int mxl_scs_parse_tclas_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req, size_t len, u8 *elems_num);
 #endif /* CONFIG_SCS_MXL */
 #endif /* MXL_WME_H */
 
diff --git a/src/common/mxl_ieee802_11_defs.h b/src/common/mxl_ieee802_11_defs.h
index 29db608fd..678625c2e 100644
--- a/src/common/mxl_ieee802_11_defs.h
+++ b/src/common/mxl_ieee802_11_defs.h
@@ -106,6 +106,13 @@ struct rnr_neighbor_ap_info {
 #define QOS_CTRL_INFO_BIT_MAP_SHIFT 9
 #define QOS_CTRL_INFO_LINKID_MASK 0xf
 #define QOS_CTRL_INFO_LINKID_SHIFT 25
+#define SCS_INTRA_ACCESS_USER_PRIO_MASK 0x7
+#define SCS_INTRA_ACCESS_ALTERNATE_QUEUE BIT(3)
+#define SCS_INTRA_ACCESS_DROP_ELIGIBILITY BIT(4)
+#define SCS_CLASSIFIER_TYPE4	4
+#define SCS_CLASSIFIER_TYPE10	10
+#define IPV4	4
+#define IPV6	6
 
 struct qos_characteristic_element{
 	u8 eid;
@@ -135,6 +142,18 @@ enum qos_variables{
 enum wfa_capabilities_element{
 	WFA_QOS_MGMT_SCS_TRAFFIC_DESC = BIT(2),
 };
+
+enum scs_classifier_type4_mask {
+	SCS_CLASS_TYPE4_VERSION = 0,
+	SCS_CLASS_TYPE4_SRC_IP = 1,
+	SCS_CLASS_TYPE4_DST_IP = 2,
+	SCS_CLASS_TYPE4_SRC_PORT = 3,
+	SCS_CLASS_TYPE4_DST_PORT = 4,
+	SCS_CLASS_TYPE4_DSCP = 5,
+	SCS_CLASS_TYPE4_PROTO = 6,
+	SCS_CLASS_TYPE4_FLOW_LABEL = 7,
+};
+
 #endif /* CONFIG_SCS_MXL */
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 #define WLAN_CAPABILITY_ML_CRITICAL_UPDATE BIT(6)
diff --git a/src/common/vendor_cmds_copy.h b/src/common/vendor_cmds_copy.h
index 1f0dd06ca..cb36cf4d7 100644
--- a/src/common/vendor_cmds_copy.h
+++ b/src/common/vendor_cmds_copy.h
@@ -425,6 +425,8 @@ enum ltq_nl80211_vendor_subcmds {
   LTQ_NL80211_VENDOR_SUBCMD_SET_PBAC,
   LTQ_NL80211_VENDOR_SUBCMD_SET_AQM_STA_EN,
   LTQ_NL80211_VENDOR_SUBCMD_GET_AQM_STA_EN,
+  LTQ_NL80211_VENDOR_SUBCMD_SET_SCS_ENABLE,
+  LTQ_NL80211_VENDOR_SUBCMD_GET_SCS_ENABLE,
   /* add new iw sub commands here */
 
   /* used to define LTQ_NL80211_VENDOR_SUBCMD_MAX below */
@@ -699,6 +701,55 @@ struct mxl_ml_sta_add_param {
   u16 aid;
 } __attribute__ ((packed));
 
+/* Filter values and filter len for the type10 classifier is obtained *
+ * by (tclas_element_ie len field - 4)/2 .since the IE len field is 1byte *
+ * so the max tclas10 filter len will be (255 - 4)/2 = 125 */
+#define MAX_TCLAS10_FILTER_LEN  125
+#define QOS_CTRL_INFO_DIR_MASK  0x3
+#define QOS_CTRL_INFO_UL        0x0
+#define QOS_CTRL_INFO_DL        0x1
+
+struct mxl_scs_tclas_info {
+  u8 tclasElemsUp;
+  u8 tclasLen;
+  u8 tclasElemstype;
+} __attribute__ ((packed));
+
+struct classIpv4Params {
+  u32 srcIp;
+  u32 dstIp;
+  u16 srcPort;
+  u16 dstPort;
+  u8 dscp;
+  u8 proto;
+} __attribute__ ((packed));
+
+struct classIpv6Params {
+  u8 srcIp[16];
+  u8 dstIp[16];
+  u16 srcPort;
+  u16 dstPort;
+  u8 dscp;
+  u8 nextHeader;
+  u8 flowLabel[3];
+} __attribute__ ((packed));
+
+struct mxl_class_type4_params {
+  u8 version;
+  union {
+    struct classIpv4Params v4;
+    struct classIpv6Params v6;
+  } u;
+} __attribute__ ((packed));
+
+struct mxl_class_type10_params {
+  u8 protoInstance;
+  u8 protoNum;
+  u8 filterLen;
+  u8 filterValue[MAX_TCLAS10_FILTER_LEN];
+  u8 filterMask[MAX_TCLAS10_FILTER_LEN];
+} __attribute__ ((packed));
+
 struct mxl_scs_add_req {
   u16 aid;
   u8  status;
@@ -720,6 +771,12 @@ struct mxl_scs_add_req {
   u8  msduDeliveryRatio;
   u8  msduCountExponent;
   u16 mediumTime;
+  u8 intraAccessPriority; //b0-b2 UP, b3= alternate queu, b4= drop eligibility, b5-b7 reserved
+  u8 version;
+  u8 direction;
+  struct mxl_scs_tclas_info tclasInfo;
+  struct mxl_class_type4_params type4Params;
+  struct mxl_class_type10_params type10Params;
 }__attribute__ ((packed));
 
 struct mxl_scs_rem_req {
-- 
2.43.0

